### DISCUSSION TITLE
Battery level on HARestAPI

### DISCUSSION THREAD

**Discussion by Olgmo** [PricelessToolkit/MailBoxGuard#34] [Q&A] [answered]
I have been using HARestAPI for a while, but don't see any battery updates.
Using a Li-Ion 14500 battery.
Is it supposed to not update battery percentage or am I missing something?

The output from serial monitor is like this:

```
`0xAAAA
-120
-12.75
Connecting: http://192.168.1.205:8123/api/services/input_boolean/turn_on
[HTTP] POST... code: 200
reply was:
==========
[{"entity_id":"input_boolean.mailbox_guard_motion","state":"on","attributes":{"editable":true,"icon":"mdi:mail","friendly_name":"mailbox_guard_motion"},"last_changed":"2023-12-12T10:09:37.236609+00:00","last_updated":"2023-12-12T10:09:37.236609+00:00","context":{"id":"01HHER1RYKBR26CV5K6W15DJTF","parent_id":null,"user_id":"acb2688cbc89415baa3ed348dd117d69"}}]
==========
closing connection
Connecting: http://192.168.1.205:8123/api/services/input_number/set_value
[HTTP] POST... code: 200
reply was:
==========
[{"entity_id":"input_number.mailbox_guard_rssi","state":"-120.0","attributes":{"initial":null,"editable":true,"min":-196.0,"max":63.0,"step":1.0,"mode":"slider","icon":"mdi:signal","friendly_name":"mailbox_guard_rssi"},"last_changed":"2023-12-12T10:09:37.787822+00:00","last_updated":"2023-12-12T10:09:37.787822+00:00","context":{"id":"01HHER1SFVCSG0RCS1BQXQ8RFH","parent_id":null,"user_id":"acb2688cbc89415baa3ed348dd117d69"}}]
==========
closing connection
Connecting: http://192.168.1.205:8123/api/services/input_number/set_value
[HTTP] POST... code: 200
reply was:
==========
[{"entity_id":"input_number.mailbox_guard_snr","state":"-12.75","attributes":{"initial":null,"editable":true,"min":-32.0,"max":32.0,"step":1.0,"mode":"slider","icon":"mdi:waves","friendly_name":"mailbox_guard_snr"},"last_changed":"2023-12-12T10:09:38.327963+00:00","last_updated":"2023-12-12T10:09:38.327963+00:00","context":{"id":"01HHER1T0QPX7VBF58JVPJND8Z","parent_id":null,"user_id":"acb2688cbc89415baa3ed348dd117d69"}}]
==========
closing connection
`
```

**Comment by PricelessToolkit** [comment]
I'll test it tomorrow and report back

**Reply by PricelessToolkit** [reply]
I didn't forget to check, I just didn't have time, I'll check in this week.

**Comment by Olgmo** [comment]
Thank you

**Reply by Olgmo** [reply]
Did som more local testing to give you some more data..
Installed the mqtt version to get some more info (mqtt_explorer).
Tried with a low voltage battery (2.6V) to test. With this battery I get BATLOW = ON (Code is 0xBBBB)
Still no battery percentage.
homeassistant/sensor/mailbox/batt is still "UNDEFINED"

I wish you a prosperous new year!

**Reply by PricelessToolkit** [reply]
This code is a contribution, I am not familiar with it. However, I think I have found the problem. 

please can you test this code? change  line 116  in "LoRa_Gateway_HARestAPI.ino"

```c
String battery_percent = String(int(round(recv.substring(battery_index + 1).toFloat())));
```

To

```c
String battery_percent = recv.substring(battery_index + 1);
```

Happy New Year!!!

**Reply by Olgmo** [reply]
Looking furthermore into this problem.
When comparing the 3 different source code for the sensor I can see that:
1: Mailbox_Guard_Sensor.ino : Has no battery percentage in the code.
2: Mailbox_Guard_Sensor_HARestAPI.ino: Has battery percentage disabled (#define TransmitBattPercent 0)
3: Mailbox_Guard_Sensor_MQTT_Auto_Discovery.ino: Has battery percentage enabled. (#define TransmitBattPercent 1)

My guess now is that the shipped sensor has is installed with INO 1 or 2. 
If this is the case it is really sad (bad) that this was not specified or an option when ordered.

Can I use a USB2TTL FT232RL to program the sensor?
If not, any Aliexpress recommandation?

**Reply by PricelessToolkit** [reply, marked as answer]
Hi, when testing the sensor, I use the "Mailbox_Guard_Sensor.ino" file. This isn't specified as the device isn't "plug-and-play"; you must select, configure, and flash the firmware. I'll update the description for clarity. For programming, any 3.3v "UPDI" or "Serial2UPDI" programmer works. If you don't have one, you can modify a 3.3v UART programmer into a Serial2UPDI using two resistors and a diode https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

on Ali I can't find any cheap UPDI Programmer, You can search at https://www.tindie.com/ or use my design "UNIProg"

**Reply by Olgmo** [reply]
Okay, Thanks for clarifying.
I made a mod-cable described here: http://silmor.de/tronic.updi.php with a 1k resistor. That worked too.
Now my "OCD" brain can relax :-)
